kratos/utilities/geometry_utilities.h

```diff
@@ -60,128 +55,131 @@ class GeometryUtils
-
-        DN_DX(0,0) = -y20 * z30 + y30 * z20 + y10 * z30 - z10 * y30 - y10 * z20 + z10 * y20;
-        DN_DX(0,1) = -z20 * x30 + x20 * z30 - x10 * z30 + z10 * x30 + x10 * z20 - z10 * x20;
-        DN_DX(0,2) = -x20 * y30 + y20 * x30 + x10 * y30 - y10 * x30 - x10 * y20 + y10 * x20;
-        DN_DX(1,0) = y20 * z30 - y30 * z20;
-        DN_DX(1,1) = z20 * x30 - x20 * z30;
-        DN_DX(1,2) = x20 * y30 - y20 * x30;
-        DN_DX(2,0) = -y10 * z30 + z10 * y30;
-        DN_DX(2,1) = x10 * z30 - z10 * x30;
-        DN_DX(2,2) = -x10 * y30 + y10 * x30;
-        DN_DX(3,0) = y10 * z20 - z10 * y20;
-        DN_DX(3,1) = -x10 * z20 + z10 * x20;
-        DN_DX(3,2) = x10 * y20 - y10 * x20;
+        const double x10 = rGeometry[1].X() - rGeometry[0].X();
+        const double y10 = rGeometry[1].Y() - rGeometry[0].Y();
+        const double z10 = rGeometry[1].Z() - rGeometry[0].Z();
+
+        const double x20 = rGeometry[2].X() - rGeometry[0].X();
+        const double y20 = rGeometry[2].Y() - rGeometry[0].Y();
+        const double z20 = rGeometry[2].Z() - rGeometry[0].Z();
+
+        const double x30 = rGeometry[3].X() - rGeometry[0].X();
+        const double y30 = rGeometry[3].Y() - rGeometry[0].Y();
+        const double z30 = rGeometry[3].Z() - rGeometry[0].Z();
+
+        const double detJ = x10 * y20 * z30 - x10 * y30 * z20 + y10 * z20 * x30 - y10 * x20 * z30 + z10 * x20 * y30 - z10 * y20 * x30;
+
+        rDN_DX(0,0) = -y20 * z30 + y30 * z20 + y10 * z30 - z10 * y30 - y10 * z20 + z10 * y20;
+        rDN_DX(0,1) = -z20 * x30 + x20 * z30 - x10 * z30 + z10 * x30 + x10 * z20 - z10 * x20;
+        rDN_DX(0,2) = -x20 * y30 + y20 * x30 + x10 * y30 - y10 * x30 - x10 * y20 + y10 * x20;
+        rDN_DX(1,0) = y20 * z30 - y30 * z20;
+        rDN_DX(1,1) = z20 * x30 - x20 * z30;
+        rDN_DX(1,2) = x20 * y30 - y20 * x30;
+        rDN_DX(2,0) = -y10 * z30 + z10 * y30;
+        rDN_DX(2,1) = x10 * z30 - z10 * x30;
+        rDN_DX(2,2) = -x10 * y30 + y10 * x30;
+        rDN_DX(3,0) = y10 * z20 - z10 * y20;
+        rDN_DX(3,1) = -x10 * z20 + z10 * x20;
+        rDN_DX(3,2) = x10 * y20 - y10 * x20;
 
-        DN_DX /= detJ;
+        rDN_DX /= detJ;
 
-        N[0] = 0.25;
-        N[1] = 0.25;
-        N[2] = 0.25;
-        N[3] = 0.25;
+        rN[0] = 0.25;
+        rN[1] = 0.25;
+        rN[2] = 0.25;
+        rN[3] = 0.25;
 
-        Volume = detJ*0.1666666666666666666667;
+        rVolume = detJ*0.1666666666666666666667;
     }
 
-    /**this function computes the element's volume (with sign)
+    /**
+     * @brief This function computes the element's volume (with sign)
      * @param rGeometry it is the array of nodes. It expects a tetrahedra
+     * @deprecated This method can be replaced by geometry function without loosing performance
      */
-    static inline double CalculateVolume3D(
-        const Element::GeometryType& rGeometry)
+    KRATOS_DEPRECATED_MESSAGE("Please use the Volume() method from the geometry")
```

Comment:
do the tet and triangle implement this efficiently, or do they use the generic version?

Reply:
The efficient one:

https://github.com/KratosMultiphysics/Kratos/blob/6b06365736c430f72eaff3ac52a0033ca39242e0/kratos/geometries/tetrahedra_3d_4.h#L404